### PR TITLE
[dag] decouple payload 2/n

### PR DESCRIPTION
**Pull request by ibalajiarun**
### Description

This PR separates Payload from (Certified) Nodes and introduces a Payload management mechanism to fetch Payloads asynchronously, like Quorum Store, but on top of DAG.

### CODE REVIEW

**Comment by ibalajiarun** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11918?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `main`
  * **PR #11362** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11362?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #11407** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11407?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #11505** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11505?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #11685** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11685?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
          * **PR #11743** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11743?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
            * **PR #11751** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11751?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
              * **PR #11808** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11808?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                * **PR #11918** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11918?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
              * **PR #11433** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11433?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11918?utm_source=stack-comment).

File: consensus/src/dag/payload/payload_fetcher.rs [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+                                    };
+                                }
+                            };
+                            let (abort_handle, abort_registration) = AbortHandle::new_pair();
+                            let abortable_fut = Abortable::new(future, abort_registration).boxed();
+                            inprogress_reqs.insert(digest, abort_handle);
+                            futures.push(abortable_fut);
+                        },
+                        FetchServiceCommand::CancelFetch(metadata) => {
+                            if let Some(handle) = inprogress_reqs.get(metadata.digest()) {
```
Comment (zekun000) [inline consensus/src/dag/payload/payload_fetcher.rs:157]: why it's get not remove?

File: consensus/src/dag/dag_driver.rs [hunk 2]
```diff
@@ -372,7 +372,20 @@ impl DagDriver {
     }
 
     pub fn try_order_all(&self) {
-        self.order_rule.lock().process_all();
+        if let Err(err) = self.order_rule.lock().process_all() {
+            self.process_order_rule_error(err);
+        }
+    }
+
+    fn process_order_rule_error(&self, err: anyhow::Error) {
```
Comment (zekun000) [inline consensus/src/dag/dag_driver.rs:380]: this seems pretty ugly, let's think about how to do it better

**Review by zekun000 — COMMENTED** [review]
did a brief pass

File: consensus/src/dag/payload/manager.rs [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+                    .1
+                    .pop_front()
+                {
+                    tx.send(txns.clone()).ok();
+                }
+                Ok(async move { rx.await }.boxed())
+            },
+            Err(DagPayloadStoreError::Missing(_)) => {
+                debug!("payload missing {}", info.id());
+                let responders = node.parents_metadata().map(|m| *m.author()).collect();
```
Comment (sasha8) [inline consensus/src/dag/payload/manager.rs:103]: Why the parents are the responders? 
Shouldn't it be the singers?

File: consensus/src/dag/payload/manager.rs [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+                        unreachable!("other payloads are not supported");
+                    };
+                    if let Some(tx) = me
+                        .waiters
+                        .remove(info.digest())
+                        .expect("must exist")
+                        .1
+                        .pop_front()
+                    {
+                        tx.send(txns.clone()).ok();
```
Comment (sasha8) [inline consensus/src/dag/payload/manager.rs:118]: I do not get this logic

File: consensus/src/dag/payload/manager.rs [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+    fn prefetch_payload(&self, node: &CertifiedNode) {
+        let DagPayload::Decoupled(info) = node.payload() else {
+            unreachable!("payload manager is only for decouple DAG payload")
+        };
+        match self.payload_store.get(info.id(), info.digest()) {
+            Ok(_) => {},
+            Err(DagPayloadStoreError::Missing(_)) => {
+                debug!("prefetch payload missing {}", node.id());
+                let responders = node.parents_metadata().map(|m| *m.author()).collect();
```
Comment (sasha8) [inline consensus/src/dag/payload/manager.rs:139]: Again, why parents?

File: consensus/src/dag/payload/manager.rs [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+pub trait TDagPayloadResolver: Send + Sync {
+    fn get_payload_if_exists(&self, node: &CertifiedNode) -> Option<Arc<DecoupledPayload>>;
+    fn add_payload(&self, payload: DecoupledPayload) -> anyhow::Result<()>;
+}
+
+pub struct DagPayloadManager {
+    dag_store: Arc<DagStore>,
+    payload_store: Arc<DagPayloadStore>,
+    requester: PayloadRequester,
+    waiters: DashMap<PayloadDigest, BoundedVecDeque<oneshot::Sender<Vec<SignedTransaction>>>>,
```
Comment (sasha8) [inline consensus/src/dag/payload/manager.rs:28]: When do you send the payloads to all senders in the queue?

File: consensus/src/dag/payload/payload_fetcher.rs [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+                            let digest = *request.payload_digest();
+
+                            if !payload_store.is_missing(id, &digest) {
+                                debug!("payload already exists: {:?}", request);
+                                continue;
+                            }
+
+                            if inprogress_reqs.contains_key(&digest) {
+                                debug!("payload already requested {:?}", request);
+                                continue;
```
Comment (sasha8) [inline consensus/src/dag/payload/payload_fetcher.rs:132]: You are droping the tx and the caller will never get the payload.
You probably should register all duplicated call and send back payload to all once received

File: consensus/src/dag/payload/manager.rs [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+            payload_store,
+            requester,
+            waiters: DashMap::new(),
+        }
+    }
+
+    pub fn insert_payload(&self, node_payload: DecoupledPayload) -> anyhow::Result<()> {
+        // Insert payload into store
+        // Cancel fetch request
+        // Notify waiters
```
Comment (zekun000) [inline consensus/src/dag/payload/manager.rs:48]: doc comments from outside?

File: consensus/src/dag/payload/manager.rs (hunk 6 above)
Comment (zekun000) [reply consensus/src/dag/payload/manager.rs:28]: why it's a bounded queue?

File: consensus/src/dag/payload/manager.rs (hunk 5 above)
Comment (zekun000) [reply consensus/src/dag/payload/manager.rs:139]: this should be the signers of the certificate?

File: consensus/src/dag/payload/manager.rs [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+        }
+    }
+}
+
+impl TDagPayloadResolver for DagPayloadManager {
+    fn get_payload_if_exists(&self, node: &CertifiedNode) -> Option<Arc<DecoupledPayload>> {
+        let DagPayload::Decoupled(info) = node.payload() else {
+            unreachable!("payload manager is only for decouple DAG payload")
+        };
+        self.payload_store.get(info.id(), info.digest()).ok()
```
Comment (zekun000) [inline consensus/src/dag/payload/manager.rs:154]: is this supposed to trigger the prefetch?

File: consensus/src/payload_manager.rs [hunk 10]
```diff
@@ -16,17 +17,60 @@ use aptos_logger::prelude::*;
...
 use futures::channel::mpsc::Sender;
 use std::sync::Arc;
-use tokio::sync::oneshot;
+use tokio::sync::{mpsc::UnboundedSender, oneshot};
 
 pub trait TPayloadManager: Send + Sync {
     fn prefetch_payload_data(&self, payload: &Payload, timestamp: u64);
 }
 
+pub struct DAGLink {
```
Comment (zekun000) [inline consensus/src/payload_manager.rs:26]: DagBatchReader?

File: consensus/src/payload_manager.rs [hunk 11]
```diff
@@ -16,17 +17,60 @@ use aptos_logger::prelude::*;
...
+            dag_payload_link_tx: tx,
+        }
+    }
+
+    async fn get_transactions(
+        &self,
+        bundle: DAGPayloadBundle,
+    ) -> ExecutorResult<Vec<SignedTransaction>> {
+        debug!("get_transactions {:?}", bundle);
+        let len = bundle.len();
```
Comment (zekun000) [inline consensus/src/payload_manager.rs:43]: this is confusing, I thought it's the number of batches, probably call it num_txns

File: consensus/src/payload_manager.rs [hunk 12]
```diff
@@ -124,10 +171,13 @@ impl PayloadManager {
                         batch_reader.clone(),
                     );
                 },
-                Payload::DirectMempool(_) => {
+                Payload::DirectMempool(_) | Payload::DAG(_) => {
                     unreachable!()
```
Comment (zekun000) [inline consensus/src/payload_manager.rs:175]: is it unreachable, do we verify the dag payload matches what we expect?

File: consensus/src/dag/payload/manager.rs [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+        node: &CertifiedNode,
+    ) -> anyhow::Result<BoxFuture<Result<Vec<SignedTransaction>, oneshot::error::RecvError>>> {
+        debug!("retrieving payload for node {}", node.id());
+        let (tx, rx) = oneshot::channel();
+        let DagPayload::Decoupled(info) = node.payload() else {
+            unreachable!("payload manager is only for decouple DAG payload")
+        };
+        self.waiters
+            .entry(*info.digest())
+            .or_insert_with(|| BoundedVecDeque::new(1))
```
Comment (zekun000) [inline consensus/src/dag/payload/manager.rs:81]: with size 1, why not just use Option?

File: consensus/src/dag/payload/payload_fetcher.rs (hunk 7 above)
Comment (zekun000) [reply consensus/src/dag/payload/payload_fetcher.rs:132]: +1 to the dropped tx here

File: consensus/src/dag/payload/payload_fetcher.rs [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+                            }
+
+                            debug!("fetching payload {:?}", request);
+
+                            let fetcher = inner.clone();
+                            let store = payload_store.clone();
+                            let future = async move {
+                                let digest = *request.payload_digest();
+                                // Fetch forever until aborted
```
[the remark this answers is not in the document]
Comment (zekun000) [reply consensus/src/dag/payload/payload_fetcher.rs:141]: in quorum store, we early return when nodes send back a proof (LI) that the data requested is already garbage collected. not sure we should loop here

File: consensus/src/dag/payload/payload_fetcher.rs [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+                else => {
+                    debug!("stopping payload fetch service");
+                    return;
+                },
+            }
+        }
+    }
+}
+
+pub struct PayloadFetcher {
```
Comment (zekun000) [inline consensus/src/dag/payload/payload_fetcher.rs:181]: there's too many similar names 🤦

File: consensus/src/dag/payload/store.rs [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+            error!("Error deleting expired payloads: {:?}", e);
+        }
+        store
+    }
+
+    pub fn prune(&self, digests: Vec<PayloadDigest>) {
+        for digest in &digests {
+            self.payload_by_digest.remove(digest);
+        }
+        if let Err(e) = self.storage.delete_certified_nodes(digests) {
```
Comment (zekun000) [inline consensus/src/dag/payload/store.rs:72]: this is not certified node lol

**Pull request closed without merging**